### PR TITLE
Specify how monolith startup should work

### LINKED ISSUE #62: Specify how monolith restarts should work
When OTT is deployed, the Monolith is restarted. Also, heroku restarts OTT about every 24 hours. In order to make this not disruptive to end users, room state is constantly being flushed to redis. When OTT starts up, it gets a list of all the rooms that were loaded and tries to restore their state.

See: https://github.com/dyc3/opentogethertube/blob/8c3b1edd6585137c0a12b2d4de4e0f3e35d1dbc4/server/roommanager.ts#L19

The problem is that if a monolith restarts, then it will load _all_ the rooms that are in redis. This results in rooms existing on more than one monolith.

constraints:
1. Can't use disk as scratch space, because disk may be wiped while monolith is offline

### PR DESCRIPTION
**Pull request by dyc3**
- update bibliography
- add monolith-startup flowchart
- add section about maintaining room state across service restarts

closes #62

### CODE REVIEW

File: final-project/manRoomStates.tex [hunk 1]
```diff
@@ -5,21 +5,50 @@ \chapter{Maintaining Room State \\
...
+
+The system needs a mechanism to detect a cold start. This can be done by having all Monoliths touch a key in redis, "monoliths\_present", every second, refreshing the expire time to 2 seconds in the future. If the key expires, then the system is in a cold start. If the system is in a cold start, monoliths that are starting up must start loading rooms from the load\_queue. Otherwise, they must not load any rooms.
+
+However, this scheme implies that "load\_queue" needs to be maintained, and invalidated if a monolith crashes. Instead, each monolith should maintain it's own set of rooms that it has loaded, with the key in the format "rooms\_loaded:UUID", with "UUID" replaced with a valid UUID. The UUID may be random, it does not need to be persistent across restarts. If the system is in a cold start, all monoliths that are starting up must merge all these sets into a new set called "load\_queue", if it does not exist. The per-monolith room sets must expire at the same time room states expire.
+
+\begin{figure}[!h]
+  \centering
+  \includegraphics[width=0.8\textwidth]{Figures/monolith-startup.png}
```
Comment (invissy) [inline final-project/manRoomStates.tex]: Not sure if it's just my laptop, but the figure is cut off on my end 
image

File: final-project/manRoomStates.tex [hunk 2]
```diff
@@ -5,21 +5,50 @@ \chapter{Maintaining Room State \\
...
+
+It's possible to simply allow the system to reach equilibrium using the mechanism described in Section \ref{Section::duplicate-rooms-across-monoliths}. However, this would result in a high memory usage on the Monolith\index{monolith} upon startup. It would also result in a lot of unnecessary network traffic and load on redis.
+
+The solution is to use redis sets\cite{redis-sets} to manage the set of rooms currently loaded. This would guarentee that each room gets loaded only once, even if the resulting distribution of rooms is not optimal.
+
+The system needs a mechanism to detect a cold start. This can be done by having all Monoliths touch a key in redis, "monoliths\_present", every second, refreshing the expire time to 2 seconds in the future. If the key expires, then the system is in a cold start. If the system is in a cold start, monoliths that are starting up must start loading rooms from the load\_queue. Otherwise, they must not load any rooms.
```
Comment (invissy) [inline final-project/manRoomStates.tex]: Is there a reason why load_queue here is not in quotation marks?

Comment (dyc3) [reply final-project/manRoomStates.tex]: No reason, ill fix it

Comment (dyc3) [reply final-project/manRoomStates.tex]: I found a better way to format these.

**Review by invissy — APPROVED** [review]
Looks good!

**Pull request merged**